cargo-clone-core/src/lib.rs

```diff
@@ -29,72 +39,81 @@ pub use cargo::{
     util::{CargoResult, Config},
 };
 
+/// Rust crate.
 #[derive(PartialEq, Eq, Debug)]
 pub struct Crate {
     name: String,
     version: Option<String>,
 }
 
 impl Crate {
+    /// Create a new [`Crate`].
+    /// If `version` is not specified, the latest version is chosen.
     pub fn new(name: String, version: Option<String>) -> Crate {
         Crate { name, version }
     }
 }
 
-pub struct CloneOpts<'a> {
-    crates: &'a [Crate],
-    srcid: &'a SourceId,
-    directory: Option<&'a str>,
-    git: bool,
+/// Clones a crate.
+pub struct Cloner {
+    /// Cargo configuration.
+    pub(crate) config: Config,
+    /// Directory where the crates will be cloned.
+    /// Each crate is cloned into a subdirectory of this directory.
+    pub(crate) directory: PathBuf,
+    /// Where the crates will be cloned from.
+    pub(crate) srcid: SourceId,
+    /// If true, use `git` to clone the git repository present in the manifest metadata.
+    pub(crate) use_git: bool,
 }
 
-impl<'a> CloneOpts<'a> {
-    pub fn new(
-        crates: &'a [Crate],
-        srcid: &'a SourceId,
-        directory: Option<&'a str>,
-        git: bool,
-    ) -> CloneOpts<'a> {
-        CloneOpts {
-            crates,
-            srcid,
-            directory,
-            git,
-        }
+impl Cloner {
+    /// Creates a new [`ClonerBuilder`] that:
+    /// - Uses crates.io as source.
+    /// - Clones the crates into the current directory.
+    pub fn builder() -> ClonerBuilder {
+        ClonerBuilder::new()
     }
-}
 
-/// Clone the specified crate from registry or git repository.
-pub fn clone(opts: &CloneOpts, config: &Config) -> CargoResult<()> {
-    let _lock = config.acquire_package_cache_lock()?;
+    /// Clone the specified crate from registry or git repository.
+    /// The crate is cloned in the directory specified by the [`ClonerBuilder`].
+    pub fn clone_in_dir(&self, crate_: &Crate) -> CargoResult<()> {
+        let _lock = self.config.acquire_package_cache_lock()?;
 
-    let multiple_crates = opts.crates.len() > 1;
-    let can_clone_in_dir = opts.directory.map(|d| d.ends_with('/')).unwrap_or(true);
-    let should_append_crate_dir = multiple_crates || can_clone_in_dir;
+        let mut src = get_source(&self.srcid, &self.config)?;
 
-    // If prefix was not supplied, use current dir.
-    let dir_path = opts
-        .directory
-        .map(PathBuf::from)
-        .unwrap_or(env::current_dir()?);
+        self.clone_in(crate_, &self.directory, &mut src)
+    }
```

Comment:
when I used cargo-clone as a library, I found confusing that he was creating or not creating the directory, based on the number of crates I passed as arguments.
So I created two separate methods.